File: Darts/src/code/ai/AbstractDartsSimulation.java
package code.ai;

import java.sql.Timestamp;

import code.listener.DartboardListener;
import code.object.Dart;
import code.screen.Dartboard;
import code.stats.GameWrapper;
import object.HandyArrayList;
import object.HashMapList;
import util.DateUtil;
import util.Debug;

public abstract class AbstractDartsSimulation implements DartboardListener
{
	protected boolean logging = false;
	
	protected AbstractDartsModel model = null;
	protected Dartboard dartboard = null;
	
	
	//Transient things
	protected Timestamp dtStart = null;
	protected Timestamp dtFinish = null;
	protected int currentRound = -1;
	protected HandyArrayList<Dart> dartsThrown = new HandyArrayList<>();
	
	protected HashMapList<Integer, Dart> hmRoundNumberToDarts = new HashMapList<>();
	
	public AbstractDartsSimulation(Dartboard dartboard, AbstractDartsModel model)
	{
		this.dartboard = dartboard;
		this.model = model;
		
		dartboard.addDartboardListener(this);
	}
	
	public abstract boolean shouldPlayCurrentRound();
	public abstract void startRound();
	public abstract int getTotalScore();
	public abstract String getGameParams();
	public abstract int getGameType();
	
	public GameWrapper simulateGame(long gameId)
	{
		resetVariables();
		
		dtStart = DateUtil.getSqlDateNow();
		
		while (shouldPlayCurrentRound())
		{
			startRound();
		}
		
		dtFinish = DateUtil.getSqlDateNow();
		
		int totalRounds = currentRound - 1;
		int totalScore = getTotalScore();
		
		Debug.appendBanner("Game Over. Rounds: " + totalRounds + ", Score: " + totalScore, logging);
		
		GameWrapper wrapper = new GameWrapper(gameId, getGameParams(), dtStart, dtFinish, totalScore);
		wrapper.setHmRoundNumberToDartsThrown(hmRoundNumberToDarts);
		wrapper.setTotalRounds(totalRounds);
		
		return wrapper;	
	}
	
	protected void resetVariables()
	{
		dartsThrown = new HandyArrayList<>();
		hmRoundNumberToDarts = new HashMapList<>();
		currentRound = 1;
	}
	protected void resetRound()
	{
		dartsThrown = new HandyArrayList<>();
		dartboard.clearDarts();
	}
	
	public void setLogging(boolean logging)
	{
		this.logging = logging;
	}
}

File: Darts/src/burlton/dartzee/code/ai/AbstractDartsSimulation.kt
package burlton.dartzee.code.ai

import burlton.dartzee.code.core.obj.HashMapList
import burlton.dartzee.code.core.util.Debug
import burlton.dartzee.code.`object`.Dart
import burlton.dartzee.code.`object`.DartsClient
import burlton.dartzee.code.db.PlayerEntity
import burlton.dartzee.code.listener.DartboardListener
import burlton.dartzee.code.screen.Dartboard
import burlton.dartzee.code.stats.GameWrapper
import burlton.dartzee.code.core.util.getSqlDateNow
import java.sql.Timestamp

abstract class AbstractDartsSimulation(protected var dartboard: Dartboard,
                                       protected var player: PlayerEntity,
                                       protected var model: AbstractDartsModel) : DartboardListener
{
    protected var logging = false

    //Transient things
    protected var dtStart: Timestamp? = null
    protected var dtFinish: Timestamp? = null
    protected var currentRound = -1
    protected var dartsThrown = mutableListOf<Dart>()

    protected var hmRoundNumberToDarts = HashMapList<Int, Dart>()

    abstract val gameParams: String
    abstract val gameType: Int

    init
    {
        dartboard.addDartboardListener(this)
    }

    abstract fun shouldPlayCurrentRound(): Boolean
    abstract fun startRound()
    abstract fun getTotalScore(): Int

    fun simulateGame(gameId: Long): GameWrapper
    {
        resetVariables()

        dtStart = getSqlDateNow()

        while (shouldPlayCurrentRound())
        {
            startRound()
        }

        dtFinish = getSqlDateNow()

        val totalRounds = currentRound - 1
        val totalScore = getTotalScore()

        Debug.appendBanner("Game Over. Rounds: $totalRounds, Score: $totalScore", logging)

        val wrapper = GameWrapper(gameId, gameParams, dtStart!!, dtFinish!!, totalScore)
        wrapper.setHmRoundNumberToDartsThrown(hmRoundNumberToDarts)
        wrapper.setTotalRounds(totalRounds)

        if (DartsClient.devMode)
        {
            wrapper.generateRealEntities(gameType, player)
        }

        return wrapper
    }

    protected open fun resetVariables()
    {
        dartsThrown = mutableListOf()
        hmRoundNumberToDarts = HashMapList()
        currentRound = 1
    }

    protected fun resetRound()
    {
        dartsThrown = mutableListOf()
        dartboard.clearDarts()
    }
}
